OCPBUGS-32776: Fix IBM Public Cloud DNS Provider Update Logic

gcs278 opened the pull request:
The IBM Public Cloud DNS provider (`cis_provider.go`) had a bug in `createOrUpdateDNSRecord` where it checked for the existence of a DNS record by filtering both DNS name and target. If the target was updated (e.g., due to a load balancer recreation), the logic would not match the existing DNS record. As a result, the function would attempt to create a new record, but fail because a record with that name already existed, as multiple DNS records with the same
name are not allowed in IBM Cloud DNS providers.

The fix is to remove the filtering by target and rely solely on filtering by name, as the name is the only attribute that needs to be unique.

This fix also includes a subtle fix in `createOrUpdateDNSRecord` of cis_provider.go to fall through to the create logic in the case of a not-found (404) error.

Additionally, the IBM DNS logic doesn't work for multiple targets and this creates unexpected and problematic results. The logic has been refactored to only create using the first target and it warns the user when multiple targets are set. This change is low risk since the Ingress Operator will never create a DNSRecord with multiple targets in `desiredDNSRecord`.

Additionally, modify TestScopeChange to check for `Available=True` after deleting the service to add test coverage for this issue.

This PR also includes some unit test fix up and missing unit test coverage for the IBM CIS Provider.

This resolves the same DNS issues for public PowerVS cloud as it uses the same logic.

gcs278 commented:
/jira refresh

gcs278 commented:
I missed one detail of the bug, missing instructions in the `Progressing` status condition for PowerVS when you change scope. Looking into it.
/hold

gcs278 commented:
@lihongan Added the missing scope change instructions for the PowerVS type.

/unhold

gcs278 commented:
infra failures
/retest

gcs278 commented:
Infra issues
/retest

lihongan commented:
pre-merge tested on OpenStack and looks good now
```
$ oc -n openshift-ingress-operator patch ingresscontroller/intlb --type=merge --patch='{"spec":{"endpointPublishingStrategy": {"type":"LoadBalancerService", "loadBalancer": {"scope":"External"}}}}'
ingresscontroller.operator.openshift.io/intlb patched

$ oc get co/ingress
NAME      VERSION                                                   AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
ingress   4.17.0-0.ci.test-2024-08-28-040831-ci-ln-13rxbgt-latest   True        True          False      23m     ingresscontroller "intlb" is progressing: IngressControllerProgressing: One or more status conditions indicate progressing: LoadBalancerProgressing=True (OperandsProgressing: One or more managed resources are progressing: The IngressController scope was changed from "Internal" to "External".  To effectuate this change, you must delete the service: `oc -n openshift-ingress delete svc/router-intlb`; the service load-balancer will then be deprovisioned and a new one created.  This will most likely cause the new load-balancer to have a different host name and IP address from the old one's.  Alternatively, you can revert the change to the IngressController: `oc -n openshift-ingress-operator patch ingresscontrollers/intlb --type=merge --patch='{"spec":{"endpointPublishingStrategy":{"loadBalancer":{"scope":"Internal"}}}}'`).

$ oc -n openshift-ingress delete svc/router-intlb
service "router-intlb" deleted

$ oc -n openshift-ingress get svc/router-intlb
NAME           TYPE           CLUSTER-IP      EXTERNAL-IP    PORT(S)                      AGE
router-intlb   LoadBalancer   172.30.110.89   38.102.83.53   80:32114/TCP,443:30834/TCP   4m23s
```
will test on IBMCloud later

lihongan commented:
pre-merge tested on IBMCloud and also looks good
```
$ oc -n openshift-ingress get svc/router-intlb
NAME           TYPE           CLUSTER-IP      EXTERNAL-IP                         PORT(S)                      AGE
router-intlb   LoadBalancer   172.30.111.21   53084b0d-eu-de.lb.appdomain.cloud   80:31026/TCP,443:32398/TCP   78s

$ oc -n openshift-ingress-operator patch ingresscontroller/intlb --type=merge --patch='{"spec":{"endpointPublishingStrategy": {"type":"LoadBalancerService", "loadBalancer": {"scope":"External"}}}}'
ingresscontroller.operator.openshift.io/intlb patched

$ oc -n openshift-ingress delete svc/router-intlb
service "router-intlb" deleted

$ oc -n openshift-ingress get svc/router-intlb
NAME           TYPE           CLUSTER-IP      EXTERNAL-IP                         PORT(S)                      AGE
router-intlb   LoadBalancer   172.30.242.68   72736508-eu-de.lb.appdomain.cloud   80:32041/TCP,443:30579/TCP   65s

$ oc -n openshift-ingress-operator get ingresscontroller/intlb -oyaml
<......>
  - lastTransitionTime: "2024-08-28T09:24:02Z"
    message: The record is provisioned in all reported zones.
    reason: NoFailedZones
    status: "True"
    type: DNSReady
```

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -794,7 +794,7 @@ func loadBalancerServiceIsProgressing(ic *operatorv1.IngressController, service
 		err := fmt.Errorf("The IngressController scope was changed from %q to %q.", haveScope, wantScope)
 		switch platform.Type {
-		case configv1.AWSPlatformType, configv1.IBMCloudPlatformType:
+		case configv1.AWSPlatformType, configv1.IBMCloudPlatformType, configv1.PowerVSPlatformType, configv1.AlibabaCloudPlatformType, configv1.OpenStackPlatformType:
```
Can this logic use `platformsWithMutableScope` so that the logic can never become inconsistent?

gcs278 replied:
Ah yes good idea. Let me look into that.

gcs278 replied:
Done.

Miciah commented:
/assign

Miciah commented:
@SzucsAti, this is a follow-up to #796.  Are you available to review the changes to the IBM DNS provider?

candita commented:
/assign

candita commented on pkg/dns/ibm/public/cis_provider.go:
```diff
@@ -161,7 +161,6 @@ func (p *Provider) createOrUpdateDNSRecord(record *iov1.DNSRecord, zone configv1
 	listOpt.SetName(dnsName)
 	for _, target := range record.Spec.Targets {
-		listOpt.SetContent(target)
 		result, response, err := dnsService.ListAllDnsRecords(listOpt)
```
Did you consider, if the target never changes, do we still need to loop on targets?  We definitely only need to make the call to `ListAllDnsRecords` once, outside the loop.

candita commented:
Please squash commits where possible.

gcs278 commented:
> Please squash commits where possible.

Ack, but I try to make my commits "one concern" or "one thing" as it makes reviews easier, it encourages specific and descriptive commit messages, it allows for more specific code reverts, and enables selective (custom) backports.

But I do agree with your expressed feeling that this PR is overloaded. How about I work with @lihongan to split up OCPBUGS-32776 into a "missing Progressing condition instructions" bug and a "IBM Public DNS Provider fix" bug, so we can have two separate PRs with these two separate concerns? I'm happy to do that to keep things cleaner, and that might help keeping backports clean.

They after all, really are separate and mostly unrelated concerns, I was just trying to conform my fixes to the current concern of the bug.

gcs278 commented:
/hold

I have some work to do per @candita [comment](https://github.com/openshift/cluster-ingress-operator/pull/1133#discussion_r1735181050), and I'd like to split this PR up, that seems like the right thing to do.

gcs278 commented:
Scope Change Instructions bug has been split into https://issues.redhat.com/browse/OCPBUGS-39151 and https://github.com/openshift/cluster-ingress-operator/pull/1135

Still working on addressing https://github.com/openshift/cluster-ingress-operator/pull/1133#discussion_r1735181050.

gcs278 commented:
/retest

lihongan commented:
/label qe-approved
pre-merged tested and passed
```
$ oc -n openshift-ingress-operator patch ingresscontroller/intlb --type=merge --patch='{"spec":{"endpointPublishingStrategy": {"type":"LoadBalancerService", "loadBalancer": {"scope":"External"}}}}'
ingresscontroller.operator.openshift.io/intlb patched

$ oc -n openshift-ingress delete svc/router-intlb
service "router-intlb" deleted

$ oc -n openshift-ingress-operator get ingresscontroller/intlb -oyaml
  - lastTransitionTime: "2024-09-04T05:35:16Z"
    message: The record is provisioned in all reported zones.
    reason: NoFailedZones
    status: "True"
    type: DNSReady

$ oc get co/ingress
NAME      VERSION                                                AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
ingress   4.18.0-0.test-2024-09-04-030153-ci-ln-41w8p6k-latest   True        False         False      94m     
```

lihongan commented:
/retest

gcs278 commented:
Sorry, I should mark WIP for now. Uncovering more unit test fixes that are needed as a result of recent updates.
/wip

gcs278 commented:
/WIP

gcs278 commented:
infra failures
/retest

gcs278 commented:
Removed WIP, sorry for the delay, I am open to reviews again @candita @Miciah @SzucsAti 

**Updates since last review:**
- **I added more unit test coverage.**
  - I felt it necessary to fix existing bugs in the unit tests and add more coverage as it provides confidence that I didn't perturb the DNS provider logic in an unintentional way.
  - The unit tests being in separate commits is intentional and provides a valuable diff before & after the *actual* bug fix here. You can review the last commit and see there were a few required unit test updates which are results of the bug fixes I added. 
- **I removed *create/update* support for multiple targets in DNSRecords for IBM Cloud.**
  - First, the code was misleading because IBM DNS doesn't support it, and therefore it doesn't work (it hot loops and overwrites records in IBM private cloud).
  - Second, since the bug fix actually impacts the way that multi-targeted DNSRecords would get created in IBM Public Cloud, it seemed logical to fix this right now, otherwise, we'd create a future headache for maintaining compatibility (if we were ever concerned about compatibility here)

gcs278 commented:
Keeping the hold while I wait to see what the results of adding a IBM Cloud job are https://github.com/openshift/release/pull/56785

/hold

gcs278 commented:
/retest

gcs278 commented:
Now that https://github.com/openshift/release/pull/56785 has merged, though it will most likely fail due to DNS Propagation issues (being tested in https://github.com/openshift/cluster-ingress-operator/pull/1132), I'm curious to see if we see anything interesting:
/test e2e-ibmcloud-operator

gcs278 commented:
I don't think it's enabled, but just curious:
/testwith e2e-ibmcloud-operator https://github.com/openshift/cluster-ingress-operator/pull/1164

gcs278 commented:
/testwith openshift/cluster-ingress-operator/master/e2e-ibmcloud-operator openshift/cluster-ingress-operator#1164

gcs278 commented:
After introducing and running e2e-ibmcloud-operator on different PRs, I realized it was passing successfully without this fix. I was mistaken that our E2E would have caught this bug. However, it looks like adding a `Availabile=True` condition check to `TestScopeChange` after deleting the service ensures test coverage for bugs like this. Updated this PR to add that.

Now let's test to make sure everything works:
/testwith openshift/cluster-ingress-operator/master/e2e-ibmcloud-operator https://github.com/openshift/cluster-ingress-operator/pull/1164

gcs278 commented:
Forgot to push...
/testwith openshift/cluster-ingress-operator/master/e2e-ibmcloud-operator https://github.com/openshift/cluster-ingress-operator/pull/1164

gcs278 commented:
/testwith openshift/cluster-ingress-operator/master/e2e-ibmcloud-operator openshift/cluster-ingress-operator/pull#1164

gcs278 commented:
/testwith openshift/cluster-ingress-operator/master/e2e-ibmcloud-operator https://github.com/openshift/cluster-ingress-operator#1164

gcs278 commented:
[multi-pr-openshift-cluster-ingress-operator-1133-openshift-cluster-ingress-operator-1164](https://github.com/openshift/cluster-ingress-operator/pull/1133/checks?check_run_id=32837360017) passed which has E2E test coverage for this bug, proving it resolved the problem.

gcs278 commented:
/retest-required

gcs278 commented:
/retest

gcs278 commented on pkg/dns/ibm/public/cis_provider.go:
```diff
@@ -160,41 +166,43 @@ func (p *Provider) createOrUpdateDNSRecord(record *iov1.DNSRecord, zone configv1
+		}
+	}
+	if result == nil || result.Result == nil {
+		return fmt.Errorf("createOrUpdateDNSRecord: ListAllDnsRecords returned nil as result")
```
Hmm, that's a good question. I found that `result` and `result.Result` are both initialized when creating the DNS record for the first time. The results gets filtered for the `record.Spec.DNSName`, so whether it's the very first DNS record in the hosted zone isn't a factor.

I agree it's a bit odd to check for `result.Result`, but the original author wrote this logic, I haven't update it as works as intended so far, and it's possible there is a subtle reason why it's here.

gcs278 reviewed (commented):
@candita thanks for the detailed review. Should be ready for your review again.

candita commented:
@gcs278 I didn't re-review this because it was on hold.  I assume that was an oversight.
/cancel hold

gcs278 commented:
> @gcs278 I didn't re-review this because it was on hold. I assume that was an oversight. /cancel hold

@candita Yea I forgot to take the hold off this time, though I often will put holds to ensure every gets a chance to review, or I'm waiting for final CI results, but still want code reviews. I could be wrong. I use WIP as an in indicator it's not ready for review.

/hold cancel

candita commented:
/retest

candita commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -190,9 +194,15 @@ func (p *Provider) createOrUpdateDNSRecord(record *iov1.DNSRecord, zone configv1
 		record.Spec.RecordTTL = defaultDNSSVCSRecordTTL
 	}
+	// We only support one target, warn the user.
+	if len(record.Spec.Targets) > 1 {
```
Why not add this to [ValidateInputDNSData](https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/dns/ibm/common.go#L88-L90)?  Shouldn't this error stop the addition of the record, rather than just logging it?

Miciah replied:
We could add a Boolean argument to `ValidateInputDNSData` for allowing/rejecting multiple targets.  That said, `ValidateInputDNSData` always felt weird to me anyway; it seems like the validation should be moved into the controller, or even enforced by API validation.  The approach in this PR seems fine to me.  Eventually, it would be nice to follow up with some rethinking and refactoring of the validation logic.

candita commented on pkg/dns/ibm/private/dnssvcs_provider.go:
```diff
@@ -201,72 +209,74 @@ func (p *Provider) createOrUpdateDNSRecord(record *iov1.DNSRecord, zone configv1
+	for _, resourceRecord := range listResult.ResourceRecords {
+		if *resourceRecord.Name == dnsName {
+			if resourceRecord.ID == nil {
+				return fmt.Errorf("createOrUpdateDNSRecord: record id is nil")
```
But why wouldn't we want to continue iterating over the `listResult.ResourceRecords` to find a match to the `dnsName`?

candita replied:
Sounds reasonable.  I was concerned that there could be a backward compatibility issue if it was possible at any time to have a missing ID.

candita commented on pkg/dns/ibm/public/cis_provider.go:
```diff
@@ -152,6 +154,10 @@ func (p *Provider) createOrUpdateDNSRecord(record *iov1.DNSRecord, zone configv1
 		log.Info("Warning: TTL must be between 120 and 2,147,483,647 seconds, or 1 for Automatic. RecordTTL set to default", "default CIS record TTL", defaultCISRecordTTL)
 		record.Spec.RecordTTL = defaultCISRecordTTL
 	}
+	// We only support one target, warn the user.
```
Same comment as https://github.com/openshift/cluster-ingress-operator/pull/1133/files#r1919281693

candita commented on pkg/dns/ibm/public/cis_provider.go:
```diff
@@ -160,41 +166,45 @@ func (p *Provider) createOrUpdateDNSRecord(record *iov1.DNSRecord, zone configv1
 	// "." when it creates a wildcard DNS record.
 	dnsName := strings.TrimSuffix(record.Spec.DNSName, ".")
 	listOpt.SetName(dnsName)
-	for _, target := range record.Spec.Targets {
-		listOpt.SetContent(target)
```
Wow, how did anything work with this here?  But do we want to add just one target to the listOpt, or is that the issue you're fixing - we don't want to match on name + target?

candita commented on pkg/dns/ibm/public/cis_provider.go:
```diff
@@ -160,41 +166,43 @@ func (p *Provider) createOrUpdateDNSRecord(record *iov1.DNSRecord, zone configv1
+
+	result, response, err := dnsService.ListAllDnsRecords(listOpt)
+	if err != nil {
+		if response == nil || response.StatusCode != http.StatusNotFound {
```
I meant, if response is nil, wouldn't that be a signal to create the record?

gcs278 replied:
Sorry for resolving this prematurely.

The response type of [DetailedResponse](https://github.com/openshift/cluster-ingress-operator/blob/10966557ef8592858fe068065e526a87c37315c5/vendor/github.com/IBM/go-sdk-core/v5/core/detailed_response.go#L27) has a `StatusCode` field. That should always be set on a valid list API call, so `response` should be not nil here.

candita commented on pkg/dns/ibm/public/cis_provider.go:
```diff
@@ -160,41 +166,45 @@ func (p *Provider) createOrUpdateDNSRecord(record *iov1.DNSRecord, zone configv1
-			return fmt.Errorf("createOrUpdateDNSRecord: ListAllDnsRecords returned nil as result")
+		log.Info("created DNS record", "record", record.Spec, "zone", zone, "target", target)
+	} else {
+		if result.Result[0].ID == nil {
```
If it's possible for `len(result.Result) > 1`, then you'd need to iterate over the results, wouldn't you?

candita commented on pkg/dns/ibm/public/cis_provider_test.go:
```diff
@@ -37,77 +42,133 @@ func Test_Delete(t *testing.T) {
 		},
 		{
-			desc:         "deleteRecordNotFound",
+			desc:         "delete failure with 404",
```
```suggestion
			desc:         "delete failure with 404 (record disappears)",
```

candita commented on pkg/dns/ibm/public/cis_provider_test.go:
```diff
@@ -37,77 +42,133 @@ func Test_Delete(t *testing.T) {
-			desc:         "listFailError",
-			DNSName:      "testDelete",
-			recordedCall: "DELETE",
+			desc:    "list failure (not 404)",
```
```suggestion
			desc:    "list failure (fake timeout)",
```

candita commented on pkg/dns/ibm/public/cis_provider_test.go:
```diff
@@ -37,77 +42,133 @@ func Test_Delete(t *testing.T) {
 		},
 		{
-			desc:         "deleteError",
+			desc:         "delete failure (not 404)",
```
```suggestion
			desc:         "delete failure (fake timeout)",
```

candita commented on pkg/dns/ibm/public/cis_provider_test.go:
```diff
@@ -171,61 +233,156 @@ func Test_createOrUpdateDNSRecord(t *testing.T) {
 		},
 		{
-			desc:         "updateError",
+			desc:         "update error",
```
```suggestion
			desc:         "update error (fake timeout)",
```

candita commented on pkg/dns/ibm/private/dnssvcs_provider_test.go:
```diff
@@ -35,82 +41,199 @@ func Test_Delete(t *testing.T) {
-				OutputStatusCode: http.StatusOK,
+		},
+		{
+			desc:    "list failure (not 404)",
```
```suggestion
			desc:    "list failure (fake timeout)",
```

candita commented on pkg/dns/ibm/private/dnssvcs_provider_test.go:
```diff
@@ -35,82 +41,199 @@ func Test_Delete(t *testing.T) {
 		},
 		{
-			desc:         "deleteError",
+			desc:         "delete failure (not 404)",
```
```suggestion
			desc:         "delete failure (fake timeout)",
```

candita commented on pkg/dns/ibm/public/cis_provider_test.go:
```diff
@@ -171,61 +233,156 @@ func Test_createOrUpdateDNSRecord(t *testing.T) {
-				OutputError:      nil,
-				OutputStatusCode: http.StatusOK,
+				OutputError:    nil,
+				OutputResponse: &core.DetailedResponse{StatusCode: http.StatusOK},
```
This test's desc should be "happy path - update" or something similar.

gcs278 reviewed (commented):
Thanks for another code review @candita! Comments should be addressed.

candita commented:
/retest

gcs278 commented:
e2e-aws-ovn-serial has issue that https://github.com/openshift/origin/pull/29447 is attempting to fix

candita commented on pkg/dns/ibm/public/cis_provider_test.go:
```diff
@@ -37,77 +41,133 @@ func Test_Delete(t *testing.T) {
+						ID:      pointer.String("testDelete"),
+						Name:    pointer.String("testDelete"),
+						Content: pointer.String("11.22.33.44"),
+					}},
```
Maybe `Expected...` but we can live with for now.

candita commented on pkg/dns/ibm/public/cis_provider_test.go:
```diff
@@ -37,77 +42,133 @@ func Test_Delete(t *testing.T) {
 		{
-			desc:         "listFailNotFound",
-			DNSName:      "testDelete",
-			recordedCall: "DELETE",
```
nit: why was this removed?

gcs278 replied:
Sorry that link isn't working for me, github is glitching trying to take me to the right line.

So `recordedCall` was removed here specifically because I changed how this test works. I saw this `listFailNotFound` as a duplicate, unnecessary, and confusing. Here it is before modification for reference: https://github.com/openshift/cluster-ingress-operator/blob/0b9e89b829ea2eb76e193a6efe9e23e303e384c8/pkg/dns/ibm/public/cis_provider_test.go#L53-L66

So it's a nil `OutputError`, meaning the API call worked, but the HTTP status code is 404 (StatusNotFound). The way the original author wrote the logic, if the `err` is nil, it doesn't even look at the status code: https://github.com/openshift/cluster-ingress-operator/blob/21971b42cc86089c823988ba62312c41491d0f2c/pkg/dns/ibm/private/dnssvcs_provider.go#L95-L99

Therefore, it's really a duplicate of the "happy path" unit test case above: https://github.com/openshift/cluster-ingress-operator/blob/0b9e89b829ea2eb76e193a6efe9e23e303e384c8/pkg/dns/ibm/public/cis_provider_test.go#L39-L52

That's why I adjusted the test to `list failure with 404`, so it would actually test a different code path. I agree that this one is really nuanced to review, and the diffs are painful to look at.

As for the other removals of `recordedCall`, they are organized within the commit: 
https://github.com/openshift/cluster-ingress-operator/pull/1133/commits/0b9e89b829ea2eb76e193a6efe9e23e303e384c8. Because they were actually incorrect. The `CallHistory` was using the `recordedCall` of the unit test previously ran (not good!). Once fixing that, a bunch of the `recordedCall`s were removed to make the unit test accurate again.

Miciah replied:
> The way the original author wrote the logic, if the `err` is nil, it doesn't even look at the status code:
[...]
> Therefore, it's really a duplicate of the "happy path" unit test case above: 
[...]
> That's why I adjusted the test to `list failure with 404`, so it would actually test a different code path.

In the spirit of black-box testing, I think it's fine to have both test cases.  Even if an examination of the logic that is being tested reveals that it uses the same code path for both cases, they are still two meaningfully different inputs from an outside perspective.  If the logic were modified, we would benefit from having both test cases to verify that both inputs continued to result in the same behavior.  

That said, don't worry about adding back the test case at this point (unless you really want to!).  I just wanted to respond to the implication that a test case that follows the same code path as another test case is necessarily redundant.

candita commented:
I think the image stream failure was fixed, so let's try:
/test e2e-aws-ovn-serial

candita commented:
/lgtm
/approve
Holding for @Miciah, who would like to look too.
/hold

gcs278 commented:
Releasing the hold due to no new reviews/comments.

/hold cancel

gcs278 commented:
While maybe not absolutely critical, I consider this fix a major fix as users are prevented from recreating IngressController's  service IBMCloud platforms (something we explicitly tell them to do when they change scope). I also have other commitments that prevent me from waiting another month to merge this.

/label acknowledge-critical-fixes-only

Miciah commented on pkg/dns/ibm/private/dnssvcs_provider_test.go:
```diff
@@ -108,6 +108,23 @@ func Test_Delete(t *testing.T) {
 			expectErrorContains: "error in ListAllDnsRecords",
 		},
+		{
+			desc:    "list success but missing ID",
```
Seems like this should be "list failure (fake timeout) and missing ID".  However, it really doesn't matter what the status is if the id is missing or mismatched, so maybe the description should be "list failure (missing ID)"?

gcs278 replied:
Good catch, a bit of a copypasta. The response code doesn't get checked if the error is nil, but I meant to just have it be a success to be a less confusing test.

Miciah replied:
Ah, I see, the HTTP error is ignored because `ListResourceRecords` returns a nil error value.  I guess "list success" refers to the error value, not the HTTP code.  It's kind of confusing to have a test case where the HTTP code indicates an error but `ListResourceRecords` reports no error.  I wouldn't have said anything if you had `http.StatusOK` instead of `http.StatusRequestTimeout`.  I guess it doesn't really matter, and the conspicuous test input did prompt me to look more closely at the logic, which is a good thing.  ¯\\\_(ツ)\_/¯

Miciah commented on pkg/dns/ibm/private/dnssvcs_provider_test.go:
```diff
@@ -35,76 +41,182 @@ func Test_Delete(t *testing.T) {
+			},
+			expectErrorContains: "delete: ListResourceRecords returned nil as result",
+		},
+		{
+			desc:    "list success but nil dns record list",
+			DNSName: "testDelete",
+			listAllDnsRecordsInputOutput: dnsclient.ListAllDnsRecordsInputOutput{
+				OutputError:    nil,
+				OutputResponse: &core.DetailedResponse{StatusCode: http.StatusOK},
+				OutputResult:   &dnssvcsv1.ListResourceRecords{},
+			},
+		},
```
The looks the same as "create happy path": https://github.com/openshift/cluster-ingress-operator/blob/c5d48ec7eeafdfa58cf005ff50d822228db65ca6/pkg/dns/ibm/private/dnssvcs_provider_test.go#L385-L400

Ah, but "list success but nil dns record list" is missing `target`.  Should it rather be called "list success but _no matching entry_ in dns record list"?

Miciah replied:
Well, I guess the "create happy path" is the same as "no matching entry in dns record list".  In both cases, the list happens to be empty, but in this case, the target is an empty string.  

It's weird that specifying an empty target doesn't cause an error, and it isn't clear that this test was intended to test that scenario.  Given that that's the current behavior, I don't mind having a test case.  However, the test case should be named clearly (something like "create with an empty target"), and we might follow-up with a fix to reject records with empty target and update the test case accordingly to expect an error.

gcs278 replied:
I think you misplaced your comment on the delete unit test, but I think you meant to comment on the one for create (same named test cases).

Yea I actually think your first thought is correct, it feels redundant. The target does not impact anything in this test. Right, I was going to say, maybe it provides value by encoding/protecting a specific behavior, but that wasn't the intention.

I'll at least update the name.

gcs278 commented:
Working through some comments
/hold

Miciah commented on pkg/dns/ibm/public/cis_provider_test.go:
```diff
@@ -37,77 +41,133 @@ func Test_Delete(t *testing.T) {
-				OutputError:      nil,
-				OutputStatusCode: http.StatusOK,
+				OutputError:    errors.New("error in ListAllDnsRecords"),
+				OutputResponse: &core.DetailedResponse{StatusCode: http.StatusRequestTimeout},
```
> This function is making sure we don't execute a delete command if there is a timeout.

Crucially, the test case also verifies that the DNS provider logic returns an error value.  This error value is what lets the DNS controller know that it needs to retry the whole operation.

Miciah commented on pkg/dns/ibm/public/cis_provider.go:
```diff
@@ -160,41 +165,45 @@ func (p *Provider) createOrUpdateDNSRecord(record *iov1.DNSRecord, zone configv1
-			if response != nil && response.StatusCode != http.StatusNotFound {
-				return fmt.Errorf("createOrUpdateDNSRecord: failed to list the dns record: %w", err)
-			}
-			continue
```
This change is really subtle, and it isn't clear to me how it relates to the description of the changes in the commit message.  (Indeed, it would be easy not to notice this change at all, but for the change to the unit tests!)

Of course, you cannot keep the literal `continue` statement since this logic is no longer inside a `for` loop.  However, you could have replaced `continue` with `return`.  Presumably, the reason you instead simply drop the `continue` is that the old logic was wrong:  It was skipping the create/update in the case of a not-found error, but the correct behavior _is_ to create in the case of a not-found error.  Am I correctly understanding the old behavior and the intent of your change?  

The commit message mentions two changes to this logic:

- "remove the filtering by target and rely solely on filtering by name" and
- "only create using the first target".

I think it makes sense to mention "fall through to the create logic in the case of a not-found error" as a distinct change as well.

gcs278 replied:
> Presumably, the reason you instead simply drop the continue is that the old logic was wrong: It was skipping the create/update in the case of a not-found error, but the correct behavior is to create in the case of a not-found error. Am I correctly understanding the old behavior and the intent of your change?

Exactly, that was my thought. One other point:
- It was inconsistent with the createOrUpdateDNSRecord in dnsvcs_provider.go for no obvious reason. Based on the reason you explained, I dropped the continue/return, reference: https://github.com/openshift/cluster-ingress-operator/blob/b0b91bb9ea12c4e09676551a0c19df4c3ce878c3/pkg/dns/ibm/private/dnssvcs_provider.go#L201-L209

I'll update the commit message.

Miciah commented:
I've finished reviewing.  Sorry for the last-minute comments!  There's nothing absolutely critical, so
/lgtm
but I can re-/lgtm if you do have time to push changes for my minor suggestions.

gcs278 commented:
@miciah ready for review/lgtm.

Miciah commented:
Thanks!
/lgtm

candita commented:
Another case where acquiring an AWS lease took 30 minutes and the test timed out after 4 hours.
/test e2e-aws-ovn-serial

gcs278 commented:
I forgot to remove the hold

/hold cancel

Installation failures and unrelated failures (nothing related to ibmcloud)
/retest
